From 87136d5807502e43dbcf4c41d4ed58338803e8b9 Mon Sep 17 00:00:00 2001
From: taosang2 <tao.sang@amd.com>
Date: Fri, 15 Dec 2023 10:38:06 -0500
Subject: [PATCH 1/5] SWDEV-435296 - Fix sporatic segment fault

Fix sporatic segment fault in texture test
via retaining image in texture object which
references the image.
The image will be released when the texture
object is destroyed.

Change-Id: Ic3fefa2d5dda6afebd1acd4d41ad310b138af6dd
---
 hipamd/src/hip_texture.cpp | 16 +++++-----------
 1 file changed, 5 insertions(+), 11 deletions(-)

diff --git a/hipamd/src/hip_texture.cpp b/hipamd/src/hip_texture.cpp
index 9eba4bd0f..55e1d0b73 100644
--- a/hipamd/src/hip_texture.cpp
+++ b/hipamd/src/hip_texture.cpp
@@ -240,8 +240,8 @@ hipError_t ihipCreateTextureObject(hipTextureObject_t* pTexObject,
       if (image == nullptr) {
         return hipErrorInvalidValue;
       }
-    } else if (image->parent()) {
-      image->retain();  // Because it will be released as a view in ihipDestroyTextureObject()
+    } else {
+      image->retain();  // will be released in ihipDestroyTextureObject()
     }
     break;
   }
@@ -281,6 +281,8 @@ hipError_t ihipCreateTextureObject(hipTextureObject_t* pTexObject,
       if (image == nullptr) {
         return hipErrorInvalidValue;
       }
+    } else {
+      image->retain();  // will be released in ihipDestroyTextureObject()
     }
     break;
   }
@@ -373,15 +375,7 @@ hipError_t ihipDestroyTextureObject(hipTextureObject_t texObject) {
     return hipErrorNotSupported;
   }
 
-  const hipResourceType type = texObject->resDesc.resType;
-  const bool isImageFromBuffer = (type == hipResourceTypeLinear) || (type == hipResourceTypePitch2D);
-  const bool isImageView = ((type == hipResourceTypeArray) || (type == hipResourceTypeMipmappedArray)) &&
-                           texObject->image->parent() != nullptr;
-  // If the texture object was created from an array, then the array owns the image SRD.
-  // Otherwise, if the texture object is a view, or was created from a buffer, then it owns the image SRD.
-  if (isImageFromBuffer || isImageView) {
-    texObject->image->release();
-  }
+  texObject->image->release();
 
   // The texture object always owns the sampler SRD.
   texObject->sampler->release();
-- 
2.43.1

